pkg/retriever/graphsyncretriever.go

```diff
@@ -23,408 +22,162 @@ import (
 }
 
-// metadataCompare compares two metadata.GraphsyncFilecoinV1s and returns true if the first is preferable to the second.
-func metadataCompare(a, b metadata.GraphsyncFilecoinV1, defaultValue bool) bool {
-	// prioritize verified deals over not verified deals
-	if a.VerifiedDeal != b.VerifiedDeal {
-		return a.VerifiedDeal
-	}
-
-	// prioritize fast retrievel over not fast retrieval
-	if a.FastRetrieval != b.FastRetrieval {
-		return a.FastRetrieval
-	}
-
-	return defaultValue
-}
+var _ TransportProtocol = &ProtocolGraphsync{}
 
-type retrievalResult struct {
-	PeerID     peer.ID
-	PhaseStart time.Time
-	Stats      *types.RetrievalStats
-	Event      *types.RetrievalEvent
-	Err        error
+type ProtocolGraphsync struct {
+	Client GraphsyncClient
 }
 
-// retrieval handles state on a per-retrieval (across multiple candidates) basis
-type graphsyncRetrieval struct {
-	*GraphSyncRetriever
-	ctx                context.Context
-	request            types.RetrievalRequest
-	eventsCallback     func(types.RetrievalEvent)
-	candidateMetadata  map[peer.ID]metadata.GraphsyncFilecoinV1
-	candidateMetdataLk sync.RWMutex
+// NewGraphsyncRetriever makes a new CandidateRetriever for Graphsync retrievals
+// (transport-graphsync-filecoinv1).
+func NewGraphsyncRetriever(getStorageProviderTimeout GetStorageProviderTimeout, client GraphsyncClient) types.CandidateRetriever {
+	return NewGraphsyncRetrieverWithConfig(getStorageProviderTimeout, client, clock.New(), 2*time.Millisecond)
 }
 
-type connectCandidate struct {
-	PeerID   peer.ID
-	Duration time.Duration
+func NewGraphsyncRetrieverWithConfig(
+	getStorageProviderTimeout GetStorageProviderTimeout,
+	client GraphsyncClient,
+	clock clock.Clock,
+	initialPause time.Duration,
+) types.CandidateRetriever {
+
+	return &parallelPeerRetriever{
+		Protocol: &ProtocolGraphsync{
+			Client: client,
+		},
+		GetStorageProviderTimeout: getStorageProviderTimeout,
+		Clock:                     clock,
+		QueueInitialPause:         initialPause,
+	}
 }
 
-type graphsyncCandidateRetrieval struct {
-	waitQueue  prioritywaitqueue.PriorityWaitQueue[connectCandidate]
-	resultChan chan retrievalResult
-	finishChan chan struct{}
+func (pg ProtocolGraphsync) Code() multicodec.Code {
+	return multicodec.TransportGraphsyncFilecoinv1
 }
 
-// RetrieveFromCandidates performs a retrieval for a given CID by querying the indexer, then
-// attempting to query all candidates and attempting to perform a full retrieval
-// from the best and fastest storage provider as the queries are received.
-func (cfg *GraphSyncRetriever) Retrieve(
-	ctx context.Context,
-	retrievalRequest types.RetrievalRequest,
-	eventsCallback func(types.RetrievalEvent),
-) types.CandidateRetrieval {
-
-	if cfg == nil {
-		cfg = &GraphSyncRetriever{}
-	}
-	if eventsCallback == nil {
-		eventsCallback = func(re types.RetrievalEvent) {}
+func (pg ProtocolGraphsync) GetMergedMetadata(cid cid.Cid, currentMetadata, newMetadata metadata.Protocol) metadata.Protocol {
```

Comment:
comment here on the semantics, as I wouldn't expect a failure to parse to get replaced with an empty metadata object